crates/mun_memory/src/diff.rs

```diff
@@ -174,117 +190,133 @@ fn append_struct_mapping(
+    // Sort in ascending order of their `myers::diff_length`.
+    myers_lengths.sort_by(
+        |LengthDescription { length: lhs, .. }, LengthDescription { length: rhs, .. }| lhs.cmp(rhs),
+    );
+
+    let mut used_deletions = vec![false; deletions.len()];
+    let mut used_insertions = vec![false; insertions.len()];
+    myers_lengths.into_iter().for_each(
+        |LengthDescription {
+             deletion_idx,
+             insertion_idx,
+             old_index,
+             new_index,
+             old_ty,
+             new_ty,
+             length,
+             old_fields,
+             new_fields,
+         }| {
+            // Skip marked fields
+            if !used_deletions[deletion_idx] && !used_insertions[insertion_idx] {
+                used_deletions[deletion_idx] = true;
+                used_insertions[insertion_idx] = true;
+
+                // If there is no difference between the old and new fields
+                mapping.push(if length == 0 {
+                    // Move the struct
+                    StructDiff::Move {
+                        old_index,
+                        new_index,
+                        old_ty,
+                        new_ty,
+                    }
+                } else {
+                    // ASSUMPTION: Don't use recursion, because all types are individually checked for
+                    // differences.
+                    // TODO: Support value struct vs heap struct?
+                    let diff = field_diff(old_fields, new_fields);
+
+                    // Edit the struct, potentially moving it in the process.
+                    StructDiff::Edit {
+                        diff,
+                        old_index,
+                        new_index,
+                        old_ty,
+                        new_ty,
+                    }
+                });
             }
-        });
-
-        // Prevent the row corresponding to the insertion entry from being used again, by inserting
-        // `std::usize::MAX`.
-        for idx in 0..num_deleted {
-            let idx = insert_idx * num_deleted + idx;
-            unsafe { *myers_lengths.get_unchecked_mut(idx) = std::usize::MAX };
-        }
-
-        // Prevent the column corresponding to the deletion entry from being used again, by
-        // inserting `std::usize::MAX`.
-        for idx in 0..num_inserted {
-            let idx = idx * num_deleted + delete_idx;
-            unsafe { *myers_lengths.get_unchecked_mut(idx) = std::usize::MAX };
-        }
-    }
+        },
+    );
 
     // Any remaining unused deletions must have been deleted.
-    for (idx, used) in used_deletions.into_iter().enumerate() {
-        if !used {
-            mapping.push(Diff::Delete {
-                index: unsafe { *deletions.get_unchecked(idx) },
-            });
-        }
-    }
+    used_deletions
+        .into_iter()
+        .zip(deletions.into_iter())
+        .for_each(|(used, (_, old_index, ty, _))| {
```

Comment:
Same here, I would write this as a regular `for` statement.

Reply:
I actually like these as you're already explicitly working with zipped iterators. I changed the other one